### PR TITLE
20240604/update pyv4 migration guide

### PR DESCRIPTION
**Pull request by databyjp**
### What's being changed:

- Update Pyv4 migration guide with further expanded explanations; bugfixes
- Update outdated (previously placeholder) imports

### Type of change:

- [x] **Documentation** updates (non-breaking change to fix/update documentation)

### How Has This Been Tested?

- [x] **Github action** – automated build completed without errors
- [x] **Local build** - the site works as expected when running `yarn start`

### CODE REVIEW

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 1]
```diff
@@ -14,7 +14,13 @@ import PythonCode from '!!raw-loader!/_includes/code/client-libraries/python_v4.
 The current Python client version is `v||site.python_client_version||`
 :::
 
-The `v4` Weaviate Python client API is very different from the `v3` API. This guide will help you understand the major changes and how to migrate your code at a high level.
+The `v4` Weaviate Python client API is a complete rewrite, aimed at an improved overall user experience. It is therefore also very different to the `v3` API, and will require re-learning of changed patterns in the way you interact with Weaviate.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:17]: (ln 17-21) Consider:
The `v4` Weaviate Python client API is a complete rewrite of the `v3` API. The changes are designed to improve developer experience by taking advantage of faster network communications and strong typing. 

If you are familiar with the `v3` API, it won't take long to learn the new API patterns. This guide highlights the major changes, and it offers suggestions on how to migrate your v3 code to the v4 API.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 2]
```diff
@@ -14,7 +14,13 @@ import PythonCode from '!!raw-loader!/_includes/code/client-libraries/python_v4.
...
+
+While this may introduce some overhead, we believe the `v4` API is a significant improvement to your developer experience. For instance, using the `v4` client will allow you to take full advantage faster speeds through the gRPC API, and additional static analysis for IDE assistance through strong typing.
+
+This guide will help you understand the major changes and how to migrate your code at a high level.
+
+Due to the extensive API surface changes, a comprehensive guide on how each function has changed is not provided here. For guidance on particular functionalities, please refer to the documentation throughout the site, [starting with these suggested sections](#how-to-migrate-your-code).
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:23]: Consider: 
This page doesn't cover every change. For sample code, see the [How-to](#list-of-howto-landing-pages). Functional changes are documented throughout the site.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 3]
```diff
@@ -96,6 +102,8 @@ To configure connection timeout values, see [Timeout values](/developers/weaviat
 </TabItem>
 </Tabs>
 
+Note that the `Client` object used in the `v3` API is [still available](./index.md#python-client-v3-api), but may be deprecated in the future.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:105]: Consider:
The `v3` style `Client` object is [still available](./index.md#python-client-v3-api), but may be deprecated in the future.

or

Note: The `v3` style `Client` object is [still available](./index.md#python-client-v3-api), but may be deprecated in the future.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 4]
```diff
@@ -106,12 +114,10 @@ The `v4` client API is very different from the `v3` API. Major user-facing chang
 
 ### Helper classes
 
-The `v4` client introduces helper classes to interact with Weaviate. These classes provide strong typing. Strong typing helps with code correctness. It also makes coding easier if your IDE has an auto-completion feature.
+The `v4` client makes extensive use of helper classes. These classes provide strong typing and thus static type checking. It also makes coding easier through your IDE's auto-completion feature.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:117]: Consider:
The `v4` client uses helper classes. These classes provide strong typing and static type checking. They work with your IDE's auto-completion feature to make coding easier.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 5]
```diff
@@ -134,25 +140,40 @@ import QuickStartCode from '!!raw-loader!/_includes/code/graphql.filters.nearTex
     language="py"
   />
 
+Note that the `Move` import exposes configuration elements for moving `near_text` search towards, or away from, the main query input.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:143]: Consider:
The `Move` import exposes configuration elements for `near_text` search that refine the distance between your query and result vectors.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 6]
```diff
@@ -134,25 +140,40 @@ import QuickStartCode from '!!raw-loader!/_includes/code/graphql.filters.nearTex
...
+
 </TabItem>
 </Tabs>
 
-The `wvc` import exposes basic types that you will need for most applications.
-
-```python
-import weaviate.classes as wvc
-```
+The `wvc` namespace exposes commonly used classes that you will need for most applications, [divided further into submodules based on their primary purpose](./index.md#helper-classes).
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:148]: Consider:

The `wvc` namespace exposes the classes that you will need for most applications. The namespace is further divided into [submodules](./index.md#helper-classes).

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 7]
```diff
@@ -134,25 +140,40 @@ import QuickStartCode from '!!raw-loader!/_includes/code/graphql.filters.nearTex
...
 
-```python
-from weaviate.collections.classes.grpc import Move
-```
 ### Interact with collections
 
-When you connect to a Weaviate database, the v4 client and the v3 client return different objects. The v3 client returns a `client` object that you interact with for CRUD and search operations. The v4 client uses a `collection` object instead.
+When you connect to a Weaviate database, the v4 API returns a `WeaviateClient` object, and the v3 API returns a `Client` object.
+
+The `v3` API's interactions were built around the `client` object (an instance of `Client`), including server interactions for CRUD and search operations.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:161]: Present tense:
The `v3` API's interactions are built around an instance of `Client` that is called the `client` object. This includes the server interactions for CRUD and search operations.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 8]
```diff
@@ -134,25 +140,40 @@ import QuickStartCode from '!!raw-loader!/_includes/code/graphql.filters.nearTex
...
+
+With the `v4` API, the main starting points for your interaction with Weaviate follow a different paradigm.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:163]: Consider:
In the `v4` API, the main starting points for your interaction with Weaviate follow a different paradigm.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 9]
```diff
@@ -177,19 +198,18 @@ import ManageDataCodeV3 from '!!raw-loader!/_includes/code/howto/manage-data.rea
   </TabItem>
 </Tabs>
 
-### Collections, classes and schemas
+### Nomenclature changes
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:201]: This title abstracts the section content too much

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 10]
```diff
@@ -177,19 +198,18 @@ import ManageDataCodeV3 from '!!raw-loader!/_includes/code/howto/manage-data.rea
...
 
-The focus on [collections](/developers/weaviate/config-refs/schema) in the v4 client is related to some corresponding changes in Weaviate Core:
+Some of the nomenclature within the Weaviate ecosystem is changing, and the client nomenclature has changed accordingly:
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:203]: Consider a more common word:

Weaviate is updating it's terminology. The client reflects the changes:

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 11]
```diff
@@ -177,19 +198,18 @@ import ManageDataCodeV3 from '!!raw-loader!/_includes/code/howto/manage-data.rea
...
 
-- Old-style "class" objects are "collections" now.
-- An old "schema" is a "collection configuration" now.
-- A "property schema" is a sub-set of the collection configuration.
+- A "class", in the sense of a set of objects, is a "collection".
+- The Weaviate "schema" can be accessed through "collection configuration".
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:206]: This doesn't distinguish schema as used with properties v. the old meaning which includes all of the collection configuration. 

(See Sebastian's notes from last week)

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 12]
```diff
@@ -177,19 +198,18 @@ import ManageDataCodeV3 from '!!raw-loader!/_includes/code/howto/manage-data.rea
...
-### Collections, classes and schemas
+### Nomenclature changes
 
-The focus on [collections](/developers/weaviate/config-refs/schema) in the v4 client is related to some corresponding changes in Weaviate Core:
+Some of the nomenclature within the Weaviate ecosystem is changing, and the client nomenclature has changed accordingly:
 
-- Old-style "class" objects are "collections" now.
-- An old "schema" is a "collection configuration" now.
-- A "property schema" is a sub-set of the collection configuration.
+- A "class", in the sense of a set of objects, is a "collection".
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:205]: This should be more specific. How should users distinguish between the terms?

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 13]
```diff
@@ -177,19 +198,18 @@ import ManageDataCodeV3 from '!!raw-loader!/_includes/code/howto/manage-data.rea
...
+- The Weaviate "schema" can be accessed through "collection configuration".
 
-There are other Weaviate Core changes as well. The new python client is designed to support these changes. Expect to find differences in the way you interact with schemas, collection configurations, and property configurations.
+Due to the architectural changes as well as evolving nomenclature, expect to find differences in the way you interact with Weaviate.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:208]: Consider:
Due to the architectural changes as well as changing terminology, expect to find differences in the way you interact with Weaviate.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 14]
```diff
@@ -204,9 +224,9 @@ You can still create a collection from a JSON definition. This may be a useful w
 
 ### Removal of builder patterns
 
-The builder patterns for constructing queries have been removed, as they could be confusing and potentially lead to invalid queries.
+The builder patterns for constructing queries have been removed, as they could be confusing, and could lead to runtime errors that would not be picked up with static analysis.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:227]: Consider:

The builder patterns for constructing queries have been removed. They could be confusing, and could lead to runtime errors that would not be picked up by static analysis.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 15]
```diff
@@ -204,9 +224,9 @@ You can still create a collection from a JSON definition. This may be a useful w
...
 
-In `v4`, queries are constructed using specific methods and its parameters.
+With the `v4` API, queries are constructed using specific methods and its parameters.
```
Comment (daveatweaviate) [inline developers/weaviate/client-libraries/python/v3_v4_migration.md:229]: Sentences shouldn't start with 'with'.

**Review by daveatweaviate — COMMENTED** [review]
Hey @databyjp thanks for the updates. I left some comments, let me know if you have any questions.

One question from me, there is a huge yarn lock file at the bottom of the PR. Is that a slip or are you updating the build environment?

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 3 above)
Comment (sebawita) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:105]: I would even go as far as "and will be deprecated in the future" :p

File: developers/weaviate/client-libraries/python/v3_v4_migration.md [hunk 16]
```diff
@@ -134,25 +140,40 @@ import QuickStartCode from '!!raw-loader!/_includes/code/graphql.filters.nearTex
...
 ### Interact with collections
 
-When you connect to a Weaviate database, the v4 client and the v3 client return different objects. The v3 client returns a `client` object that you interact with for CRUD and search operations. The v4 client uses a `collection` object instead.
+When you connect to a Weaviate database, the v4 API returns a `WeaviateClient` object, and the v3 API returns a `Client` object.
+
+The `v3` API's interactions were built around the `client` object (an instance of `Client`), including server interactions for CRUD and search operations.
+
+With the `v4` API, the main starting points for your interaction with Weaviate follow a different paradigm.
+
+Server-level interactions such as checking readiness (`client.is_ready()`) or getting node statuses (`client.cluster.nodes()`) still remain with `client` (now an instance of `WeaviateClient`). Interactions with CRUD and search operations, however, are now done with a `Collection` object instead, to reflect that almost all such operations target a particular collection.
```
[the remark this answers is not in the document]
Comment (sebawita) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:165]: Hey @daveatweaviate, what do you mean by `(Can you drop 'usually'?)` ?
There is no "usually" word anywhere on this page, only in the new line you proposed 🤔

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 11 above)
Comment (sebawita) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:206]: Actually, neither approach explains the topic correctly.

how about something like this:
> - A "Class" is now called a "Collection". A collection stores data objects together with their vector embeddings.
> - A "Schema" is now called a "Collection Configuration", a set of settings that define collection name, used vectorizers, index configuration, property defintions, etc.
> - We introduced "Property Schema", which refers to the definition of property names, their datatypes and other property configurations.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (daveatweaviate) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:165]: That's true, "usually" replaces "almost all." It woudl be better to drop both of them.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 4 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:117]: I was looking to highlight the extent to which they are used.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 1 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:17]: I'm very reluctant to suggest how long it should take users to learn new things. It's a complete rewrite and basically every line of code will need to be re-written.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 5 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:143]: > If it is the former (which I believe should be the case) then we should focus on explaining that you can import each class, and skip the explanation of what Move is used for.

Yup agreed

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:165]: X-refs involve to multiple collections, although there is a from/to distinction.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 11 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:206]: Fair. I consider this to be something that deserves a site-wide review, with a coordinated change and some dev education. Perhaps worth tackling in a separate PR.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 2 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:23]: Updated

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 3 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:105]: Updated

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 5 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:143]: Done

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 14 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:227]: Updated

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 6 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:148]: Updated

**Comment by databyjp** [comment]
> Hey @databyjp thanks for the updates. I left some comments, let me know if you have any questions.
> 
> One question from me, there is a huge yarn lock file at the bottom of the PR. Is that a slip or are you updating the build environment?

@daveatweaviate I believe it's best practice to update these ([source](https://yarnpkg.com/getting-started/qa#should-lockfiles-be-committed-to-the-repository)). If the plan is to not commit it, we should add it to the gitignore.

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 13 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:208]: Updated

File: developers/weaviate/client-libraries/python/v3_v4_migration.md (hunk 11 above)
Comment (databyjp) [reply developers/weaviate/client-libraries/python/v3_v4_migration.md:206]: I am going to leave out "property schema" as that bigram isn't used anywhere else.

**Pull request merged**